repositories/shopRepository.go

```diff
@@ -0,0 +1,219 @@
+	GetShop(string) (models.Shop, error)
+	AddShop(*models.Shop) error
+	UpdateShop(models.Shop) error
+	DeleteShop(string) error
+}
+
+type shopRepository struct {
+	DB *pgxpool.Pool
+}
+
+func NewShopRepo(db *pgxpool.Pool) ShopRepository {
+	if db == nil {
+		panic("ShopRepository not created, pgxpool is nil")
+	}
+	return &shopRepository {
+		DB: db,
+	}
+}
+
+func (repository *shopRepository) GetAllShops() ([]models.Shop, error) {
+	shops := []models.Shop{}
+	rows, err := repository.DB.Query(context.Background(), "select * from public.shops")
+	defer rows.Close()
+
+	if err != nil {
+		return nil, err
+	}
+
+	for rows.Next() {
+		var shop persistence.Shops
+		shop.Scan(&rows)
+
+		var stringUUID_ID string
+		err := shop.Id.AssignTo(&stringUUID_ID)
+		if err != nil {
+			return shops, err
+		}
+
+		var stringUUID_IDC string
+		err = shop.Idc.AssignTo(&stringUUID_IDC)
+		if err != nil {
+			return shops, err
+		}
+
+		shops = append(shops, models.Shop{
+			ID:     stringUUID_ID,
+			Name:	shop.Name,
+			IDC:	stringUUID_IDC,
+			Lat:	shop.Lat,
+			Lon:	shop.Lon,
+		})
+	}
+	return shops, nil
+}
+
+func (repository *shopRepository) GetShopsByLatLon(lat, lon float64) ([]models.Shop, error) {
+	shops := []models.Shop{}
+	rows, err := repository.DB.Query(context.Background(), "select * from public.shops where lat=$1 and lon=$2", lat, lon)
+	defer rows.Close()
+
+	if err != nil {
+		return nil, err
+	}
+
+	for rows.Next() {
+		var shop persistence.Shops
+		shop.Scan(&rows)
+
+		var stringUUID_ID string
+		err := shop.Id.AssignTo(&stringUUID_ID)
+		if err != nil {
+			return shops, err
+		}
+
+		var stringUUID_IDC string
+		err = shop.Idc.AssignTo(&stringUUID_IDC)
+		if err != nil {
+			return shops, err
+		}
+
+		shops = append(shops, models.Shop{
+			ID:     stringUUID_ID,
+			Name:	shop.Name,
+			IDC:	stringUUID_IDC,
+			Lat:	shop.Lat,
+			Lon:	shop.Lon,
+		})
+	}
+	return shops, nil
+}
+
+func (repository *shopRepository) GetShop(id string) (models.Shop, error) {
+	var shop models.Shop
+
+	rows, err := repository.DB.Query(context.Background(), `select * from public.shops where id = $1`, id)
+	defer rows.Close()
+
+	if err != nil {
+		return shop, err
+	}
+
+	if !rows.Next() {
+		return shop, utils.NoDataError
+	}
+
+	var shopPers persistence.Shops
+	shopPers.Scan(&rows)
+
+	var stringUUID_ID string
+	err = shopPers.Id.AssignTo(&stringUUID_ID)
+	if err != nil {
+		return shop, err
+	}
+
+	var stringUUID_IDC string
+	err = shopPers.Idc.AssignTo(&stringUUID_IDC)
+	if err != nil {
+		return shop, err
+	}
+
+	shop = models.Shop{
+		ID:     stringUUID_ID,
+		Name:	shopPers.Name,
+		IDC:	stringUUID_IDC,
+		Lat:	shopPers.Lat,
+		Lon:	shopPers.Lon,
+	}
+	return shop, nil
+}
+
+func (repository *shopRepository) AddShop(shop *models.Shop) error {
+	tx, err := repository.DB.Begin(context.Background())
+	if err != nil {
+		return err
+	}
+
+	defer tx.Rollback(context.Background())
+
+	shop.ID = uuid.NewV4().String()
```

Comment:
If you're using a pointer as a parameter you must ALWAYS check if it's nil, since this could segfault.